File: wbcorr/compute4thOrderMoments.c
void compute4thOrderMoments(double* moments, double* Z, int* p, int* n) {
    int a = 0;
    for (int i = 0; i < *p; i++) {
    	for (int j = 0; j <= i; j++) {
    	    for (int k = 0; k <= j; k++) {
    		for (int h = 0; h <= k; h++) {
    		    for (int b = 0; b < *n; b++) {
    			moments[a] = moments[a] +
    			    *(Z + (*n * i + b)) *
    			    *(Z + (*n * j + b)) *
    			    *(Z + (*n * k + b)) *
    			    *(Z + (*n * h + b));
    		    }
		    a++;
    		}
    	    }
    	}
    }
}

File: wbcorr/compute4thOrderMoments.R
compute4thOrderMoments <-function( data, jj ) {

    X <- data[[jj]]

    n <- nrow(X)
    p <- ncol(X)
    means <- apply(X, 2, mean)
    sds <- apply(X, 2, sd)
    Z <- sweep(X, 2, means, "-")
    Z <- sweep(Z, 2, sds, "/")

    q <- p*(p+1)*(p+2)*(p+3)/24
    moments <- rep(0, q)

    moments <- .C("compute4thOrderMoments",
                moments = moments,
                Z = Z,
                p = as.integer(p),
                n = as.integer(n))$moments

    moments / (n - 1)

}
